docs/sources/reference/api/docker_remote_api_v1.17.md

```diff
@@ -802,6 +804,138 @@ Status Codes:
 -   **404** – no such container
 -   **500** – server error
 
+### Get files or folders from a container as a Tar Archive.
+
+`GET /containers/(id)/copy`
+
+Get an archive of a resource in the filesystem of container `id`.
+
+Query Parameters:
+
+-   **path** - The resource in the container's filesystem. If not an absolute
+        path, it is relative to the container's working directory. If path is
+        inside the mount point of a volume, the volume's data is archived. If
+        the path specifies a directory and a volume is mounted lower in that
+        directory hierarchy, the volume's contents are not included in the
+        archive. Required.
+
+**Example request**:
+
+        GET /containers/8cce319429b2/copy?path=/foo/ HTTP/1.1
+
+**Example response**:
+
+        HTTP/1.1 200 OK
+        Content-Type: application/x-tar
+
+        {{ TAR STREAM }}
+
+Status Codes:
+
+-   **200** – no error
+-   **400** - client error, details in JSON response body, one of:
+        - path cannot be empty (missing path query parameter)
+        - no such file or directory (path does not exist)
+        - not a directory (path ends with a separator but exists as a file)
+-   **404** – no such container
+-   **500** – server error
+
+### Upload files or folders to a container from a Tar Archive.
+
+`PUT /containers/(id)/copy`
+
+Upload an archive of a resource to be extracted to a path in the filesystem
+of container `id`.
+
+Query Parameters:
+
+-   **srcPath** - Specifies the original path from which the payload archive
+        was created. Used as a hint to the server when determining if the
+        contents of the archive are a single file or should be interpreted
+        as directory contents. If the first entry of the archive has a name
```

Comment:
This is critical in solving the issue from #4272. It really only needs to be included if the source was a file and not a directory or directory contents and doesn't need to be the whole path, only the basename. This is used in https://github.com/jlhawn/docker/commit/5168da193e25b6a8e3d86b564c08fd7610a19513#diff-d50c86374d8a7d3552ed2ae1b708a7d4R97 to determine whether the given archive is just a file or is directory contents. Read that to see how it first wraps the archive to attempt to detect if the first entry of the archive is a file and matches the basename of the source path.

Why not just tell this function that it's a file? When you do `docker cp container:myfile.txt ./` we have no clean way of getting back information about whether `myfile.txt` was a file or was a directory. We can't put it in the payload of the API response because the payload is a Tar archive. I tried using an HTTP header to convey this but the job that returns the tar archive writes directly to the `http.ResponseWriter` object as stdout of the job and there's no way to set a header from inside the job and once it starts writing the response you can't go back and change the header. It is indeed unfortunate but we've tried other approaches and there seems to be no way to convey this without either specifying an option to `docker cp` as in #8362 or having a separate API endpoint to return file info for a path with the latter option leaving a race condition.